src/osd/scrubber/scrub_reservations.h

```diff
@@ -0,0 +1,125 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+#pragma once
+
+#include <cassert>
+#include <chrono>
+#include <memory>
+#include <optional>
+#include <set>
+#include <span>
+#include <string>
+#include <string_view>
+#include <vector>
+
+#include "osd_scrub_sched.h"
+
+namespace Scrub {
+
+/**
+ * Reserving/freeing scrub resources at the replicas.
+ *
+ * When constructed - sends reservation requests to the acting_set OSDs, one
+ * by one.
+ * Once a replica's OSD replies with a 'grant'ed reservation, we send a
+ * reservation request to the next replica.
+ * A rejection triggers a "couldn't acquire the replicas' scrub resources"
+ * event. All granted reservations are released.
+ *
+ * Reserved replicas should be released at the end of the scrub session. The
+ * one exception is if the scrub terminates upon an interval change. In that
+ * scenario - the replicas discard their reservations on their own accord
+ * when noticing the change in interval, and there is no need (and no
+ * guaranteed way) to send them the release message.
+ *
+ * Timeouts:
+ *
+ *  Slow-Secondary Warning: (not re-implemented yet \todo)
+ *
+ *  Reservation Timeout:
+ *  We limit the total time we wait for the replicas to respond to the
+ *  reservation request. If the reservation back-and-forth does not complete
+ *  within <conf> milliseconds, we give up and release all the reservations
+ *  that have been acquired until that moment.
+ *  (Why? because we have encountered instances where a reservation request was
+ *  lost - either due to a bug or due to a network issue.)
+ */
+class ReplicaReservations {
+  using clock = std::chrono::system_clock;
+  using tpoint_t = std::chrono::time_point<clock>;
+  using replica_subset_t = std::span<const pg_shard_t>;
+
+  ScrubMachineListener& m_scrubber;
+  PG* m_pg;
+  CephContext* m_cct;
+
+  const pg_shard_t m_whoami;
+  const spg_t m_pgid;
+
+  /// used to send both internal and inter-OSD messages
+  OSDService* m_osds;
+
+  const ConfigProxy& m_conf;
+
+  /// the acting set (not including myself), sorted by OSD id
+  std::vector<pg_shard_t> m_sorted_secondaries;
+
+  /// the next replica to which we will send a reservation request
+  std::vector<pg_shard_t>::const_iterator m_next_to_request;
+
+  /// the number of secondaries
+  size_t m_total_needeed{0};
+
+  /// always <= m_total_needed
+  size_t m_requests_sent{0};
+
+  tpoint_t m_request_sent_at;  ///< for detecting slow peers
+
+  std::string m_log_msg_prefix;
```

Comment:
No.  The way you are constructing this message in ReplicaReservations::ReplicaReservations(...)

```
  m_log_msg_prefix = fmt::format(
      "osd.{} ep: {} scrubber::ReplicaReservations pg[{}]: ", m_whoami.osd,
      epoch, m_pgid);
```

will cause it to have a different prefix from other pg log lines so it won't be grep-able and will not capture the normal PG state I, for one, consider super useful to have on each log line.

You are already capturing a pointer to the actual PG with its own gen_prefix() implementation.  Forward to it as you do with PgScrubber::gen_prefix.